README.md
# docx2python

Extract docx headers, footers, text, footnotes, endnotes, properties, and images to a Python object.

## Note to Users / Contributors

I will be doing very little coding in 2022. I will address "show stopper" bugs in docx2python, and I will accept pull requests *if* they are complete with
* an example `*.docx` file showing the problem addressed or missing feature
* a new test file with (Pytest) tests for the new fix or feature

`README_DOCX_FILE_STRUCTURE.md` may help if you'd like to extend docx2python.

## Back to docx2python

For a summary of what's new in docx2python 2, scroll down to **New in docx2python Version 2**

The code is an expansion/contraction of [python-docx2txt](https://github.com/ankushshah89/python-docx2txt) (Copyright (c) 2015 Ankush Shah). The original code is mostly gone, but some of the bones may still be here.

__shared features__:
* extracts text from docx files
* extracts images from docx files

__additions:__
* extracts footnotes and endnotes
* converts bullets and numbered lists to ascii with indentation
* converts hyperlinks to ``<a href="http:/...">link text</a>``
* retains some structure of the original file (more below)
* extracts document properties (creator, lastModifiedBy, etc.)
* inserts image placeholders in text (``'----image1.jpg----'``)
* inserts plain text footnote and endnote references in text (``'----footnote1----'``)
* (optionally) retains font size, font color, bold, italics, and underscore as html
* extract math equations
* extract user selections from checkboxes and dropdown menus
  
__subtractions:__
* no command-line interface
* will only work with Python 3.7+


## Installation
```bash
pip install docx2python
```

## Use

```python
from docx2python import docx2python

# extract docx content
docx2python('path/to/file.docx')

# extract docx content, write images to image_directory
docx2python('path/to/file.docx', 'path/to/image_directory')

# extract docx content, ignore images
docx2python('path/to/file.docx', extract_image=False)

# extract docx content with basic font styles converted to html
docx2python('path/to/file.docx', html=True)
```

Note on html feature:
* supports ``<i>``italic, ``<b>``bold, ``<u>``underline, ``<s>``strike, ``<sup>``superscript, ``<sub>``subscript, ``<span style="font-variant: small-caps">``small caps, ``<span style="text-transform:uppercase">``all caps, ``<span style="background-color: yellow">``highlighted, ``<span style="font-size:32">``font size, ``<span style="color:#ff0000">``colored text.
* hyperlinks will always be exported as html (``<a href="http:/...">link text</a>``), even if ``html=False``, because I couldn't think of a more canonical representation.
* every tag open in a paragraph will be closed in that paragraph (and, where appropriate, reopened in the next paragraph). If two subsequenct paragraphs are bold, they will be returned as `<b>paragraph a</b>`, `<b>paragraph b</b>`. This is intentional to make  each paragraph its own entity. 
* if you specify `html=True`, `&`, `>` and `<` in your docx text will be encoded as `&amp`, `&gt;` and `&lt;`

## Return Value

Function `docx2python` returns an object with several attributes.

__header__ - contents of the docx headers in the return format described herein

__footer__ - contents of the docx footers in the return format described herein

__body__ - contents of the docx in the return format described herein

__footnotes__ - contents of the docx in the return format described herein

__endnotes__ - contents of the docx in the return format described herein

__document__ - header  + body + footer (read only)

__text__ - all docx text as one string, similar to what you'd get from `python-docx2txt`

__properties__ - docx property names mapped to values (e.g., `{"lastModifiedBy": "Shay Hill"}`)

__images__ - image names mapped to images in binary format. Write to filesystem with

```
for name, image in result.images.items():
    with open(name, 'wb') as image_destination:
        write(image_destination, image)
```

__docx_reader__ - a DocxReader (see `docx_reader.py`) instance with several methods for extracting xml portions.


## Return Format

Some structure will be maintained. Text will be returned in a nested list, with paragraphs always at depth 4 (i.e., `output.body[i][j][k][l]` will be a paragraph).

If your docx has no tables, output.body will appear as one a table with all content in one cell:

```python
[  # document
    [  # table
        [  # row
            [  # cell
                "Paragraph 1",
                "Paragraph 2",
                "-- bulleted list",
                "-- continuing bulleted list",
                "1)  numbered list",
                "2)  continuing numbered list"
                "    a)  sublist",
                "        i)  sublist of sublist",
                "3)  keeps track of indention levels",
                "    a)  resets sublist counters"
            ]
        ]
     ]
 ]
```

Table cells will appear as table cells. Text outside tables will appear as table cells.


A docx document can be tables within tables within tables. Docx2Python flattens most of this to more easily navigate
within the content. 
    
## Working with output

This package provides several documented helper functions in [the ``docx2python.iterators`` module](https://docx2python.readthedocs.io/en/latest/docx2python.html#module-iterators). Here are a few recipes possible with these functions:

```python
from docx2python.iterators import enum_cells

def remove_empty_paragraphs(tables):
    for (i, j, k), cell in enum_cells(tables):
        tables[i][j][k] = [x for x in cell if x]
```

```
>>> tables = [[[['a', 'b'], ['a', '', 'd', '']]]]
>>> remove_empty_paragraphs(tables)
    [[[['a', 'b'], ['a', 'd']]]]
```

```python
from docx2python.iterators import enum_at_depth

def html_map(tables) -> str:
    """Create an HTML map of document contents.

    Render this in a browser to visually search for data.
    
    :tables: value could come from, e.g.,
        * docx_to_text_output.document
        * docx_to_text_output.body
    """

    # prepend index tuple to each paragraph
    for (i, j, k, l), paragraph in enum_at_depth(tables, 4):
        tables[i][j][k][l] = " ".join([str((i, j, k, l)), paragraph])

    # wrap each paragraph in <pre> tags
    for (i, j, k), cell in enum_at_depth(tables, 3):
        tables[i][j][k] = "".join(["<pre>{x}</pre>".format(x) for x in cell])

    # wrap each cell in <td> tags
    for (i, j), row in enum_at_depth(tables, 2):
        tables[i][j] = "".join(["<td>{x}</td>".format(x) for x in row])

    # wrap each row in <tr> tags
    for (i,), table in enum_at_depth(tables, 1):
        tables[i] = "".join("<tr>{x}</tr>".format(x) for x in table)

    # wrap each table in <table> tags
    tables = "".join(['<table border="1">{x}</table>'.format(x) for x in tables])

    return ["<html><body>"] + tables + ["</body></html>"]
```

```
>>> tables = [[[['a', 'b'], ['a', 'd']]]]
>>> html_map(tables)
<html>
    <body>
        <table border="1">
            <tr>
                <td>
                    '(0, 0, 0, 0) a'
                    '(0, 0, 0, 1) b'
                </td>
                <td>
                    '(0, 0, 1, 0) a'
                    '(0, 0, 1, 1) d'
                </td>
            </tr>
        </table>
    </body>
</html>
```

[See helper functions.](https://docx2python.readthedocs.io/en/latest/index.html)

Some fine print about checkboxes:

MS Word has checkboxes that can be checked any time, and others that can only be checked when the form is locked.
The previous print as. ``\u2610`` (open checkbox) or ``\u2612`` (crossed checkbox). Which this module, the latter will
too. I gave checkboxes a bailout value of ``----checkbox failed----`` if the xml doesn't look like I expect it to,
because I don't have several-thousand test files with checkboxes (as I did with most of the other form elements).
Checkboxes *should* work, but please let me know if you encounter any that do not.

# New in docx2python Version 2

## merge consecutive runs with identical formatting

MS Word will break up text runs arbitrarily, often in the middle of a word.


    <w:r>
        <w:t>work to im</w:t>
    </w:r>
    <w:r>
        <w:t>prove docx2python</w:t>
    </w:r>

This makes things like algorithmic search-and-replace problematic. Docx2python does not currently write docx files,
but I often use docx templates with placeholders (e.g., `#CATEGORY_NAME#`) then replace those placeholders with data.
This won't work if your placeholders are broken up (e.g, `#CAT`, `E`, `GORY_NAME#`).

Docx2python v1 merges such runs together when exporting text. Docx2python v2 will merge such runs in the XML as a
pre-processing step. This will allow saving such "repaired" XML later on.

## merge consecutive links with identical hrefs

MS Word will break up links, giving each link a different `rId`, even when these `rIds` point to the same address.

    <w:hyperlink r:id="rId13">  # rID13 points to https://github.com/ShayHill/docx2python
        <w:r>
            <w:t>docx2py</w:t>
        </w:r>
    </w:hyperlink>
    <w:hyperlink r:id="rId14">  # rID14 ALSO points to https://github.com/ShayHill/docx2python
        <w:r>
            <w:t>thon</w:t>
        </w:r>
    </w:hyperlink>

This is similar to the broken-up runs, but the cause is a little deeper in. Docx2python v1 makes a mess of these.

    <a href="https://github.com/ShayHill/docx2python">docx2py</a>
    <a href="https://github.com/ShayHill/docx2python">thon</a>

Docx2python v2 will merge such links together in the XML as a pre-processing step. As above, this will allow saving
such "repaired" XML later on.

## correctly handle nested paragraphs

MS Word will nest paragraphs

	<w:p>
		<w:r>
			<w:t>text</w:t>
		</w:r>
		<w:p>  # paragraph inside a paragraph
			<w:r>
				<w:t>text</w:t>
			</w:r>
		</w:p>
		<w:r>
			<w:t>text</w:t>
		</w:r>
	</w:p>

I haven't been able to create such a paragraph, but I've found a few files that have them. Docx2pyhon v1 will omit
closing html tags when a new paragraph is opened before the old paragraph is closed.

    <b>outer par bold text

    <i>This text is in nested par (not bold)</i>

    outer par bold text</b>

Docx2python v2 will correctly handle such cases, but this will require substantial internal changes to the way
docx2python opens and closes paragraphs.

    <b>outer par bold text</b>

    <i>This text is in nested par (not bold)</i>

    </b>outer par bold text</b>

## paragraph styles

The internal changes allow for easy access to paragraph styles (e.g., `Heading 1`). Docx2python v1 ignores these, even
with `html=True`. Docx2python v2 will capture paragraph styles.

    <h1>h1 is a paragraph style<b>bold is a run style</b></h1>

## export xml

To allow above-described light editing (e.g., search and replace), docx2python v2 will give the user access to

    1. extracted xml files
    2. the functions used to write these files to a docx

The user can only go so far with this. A docx file is built from folders full of xml files. None of these xml
files are self contained. But search and replace is enough to make document templates (documents with placeholders for
data), and that's pretty useful in itself.

## expose some intermediate functionality

Navigating through XML is straightforward with `xml.etree.ElementTree`. It is a separate step to take whatever you find and bring it
*out* of the XML. For instance, you may want to iterate over a document, looking for paragraphs with a particular
format, then pull the text out of those paragraphs. Docx2python v1 did not separate or expose "iter the document" and
"pull the content". Docx2python v2 separates and exposes these steps. This will allow easier extension.

See the `docx_reader.py` module and simple examples in the `utilities.py` module.

## see utilities.py for examples of major new features.


changelog.md
---- version 1.25 - 200820 Added support for Table of Contents text

A docx table of contents is built like a set of hyperlinks, with each hyperlink element's having an anchor (internal link) instead of an href (external link).

Previously any document with a Table of Contents would fail with `KeyError: '{http://schemas.openxmlformats.org/officeDocument/2006/relationships}id'` after failing to find an href. Now, docx2python will continue without warning if an href is not found in a hyperlink element. In an href is found, docx2python will print the href inside `<a href="{}">` as before. Anchor (internal link) elements are meaningless outside the docx and are therefore ignored by docx2python.


---- version 1.26 - 201005 Continue (with bullet) when numbering-format lookup fails

Some documents created in Pages use a different indexing scheme to specify numbered-list formats and values. I cannot infer formats or values from such files without potentially changing existing behavior. The previous behavior in such cases was to fail with an IndexError. v1.26 will now replace any numbering format with a "bullet" (--) when the format or value cannot be inferred.

This will only happen where the program would previously have failed with an IndexError, so no previous behavior (which allowed the program to complete) has been altered.


---- version 1.27 - 201102 Continue when document properties are not found

`docx2python(file).properties` returns a dictionary of document properties (e.g., {'Author': 'Shay Hill'}). Google Docs (and perhaps others) do not store such properties. When document properties cannot be found, v1.27 will continue and return an empty dictionary for `docx2python(file).properties`.

This will only happen where the program would previously have failed with a KeyError, so no previous behavior (which allowed the program to complete) has been altered.


---- version 1.27.1 - 201115 Continue when image r:id is not found

A user found a docx `imagedata` element with a missing `r:id` element. The `r:id` number gives the location of an image filename. I presume this `imagedata` element is a vector graphic, which `docx2python` does not and will not support. This makes two out of three `r:id` lookup positions (`hyperlink`, `image`, and `imagedata`) for which users have found absent `r:id`. None so far have contained anything meaningful for text export (internal links in a previous case and vector graphics in this case). Now all `r:id` lookups take place within `suppress(KeyId)` context.

This will only happen where the program would previously have failed with a KeyError, so no previous behavior (which allowed the program to complete) has been altered.


---- version 2.0.0 - big changes

--  Join run elements internally when docx2python cannot differentiate style.

If you've ever unzipped a docx file and searched for a word in your document, you probably didn't find it. This is because MSWord splits continuous text into smaller runs if the runs differ in spell-check accuracy, revision time, and other characteristics docx2python does not extract. This makes it hard to, for instance, search and replace text in the xml. Docx2Python v2 reads through the xml and joins such runs as a pre-processing step. This greatly simplifies searching output for formatted text. This will allow search and replace and other light xml operations in the future. Runs with different formatting are not joined, even if html=False is set.

--  Return text split into paragraphs (as previous version) or runs (new to Docx2Python v2).

The previous header, footer, body, footnotes, and endnotes attributes returned docx content as a 4-deep nested list of paragraph text. (paragraphs as strings): ``[[[["This is a paragraph"]]]]``. These attributes are still available. New attributes header_runs, footer_runs, etc. return docx content as a 5-deep nested list of run strings (paragraphs as lists of strings): ``[[[[['This' , ' is a ', 'paragraph']]]]]``

--  No more nested HTML styles.

Docx2Python v1 would simplify html tags: ``<b>bold text <i>bold-italic</i> more bold text</b>``. This makes an attractive export, but complicates searching / filtering for formatted text.

Docx2Python v2 will not nest html tags: ``<b>bold text </b><b><i>bold-italic</i></b><b> more bold text</b>``.

``_runs`` attributes will return ``[<b>"bold text </b>", "<b><i>bold-italic</i></b>", "<b> more bold text</b>]``.

--  More html run styles.

Now supports ``<i>``italic, ``<b>``bold, ``<u>``underline, ``<s>``strike, ``<sup>``superscript, ``<sub>``subscript, ``<span style="font-variant: small-caps">``small caps, ``<span style="text-transform:uppercase">``all caps, ``<span style="background-color: yellow">``highlighted, ``<span style="font-size:32">``font size, ``<span style="color:#ff0000">``colored text.

This is extensible. Styles can be added and removed. Note that the style change for font size has been updated from ``<font size="32">`` to ``<span style="font_size:32">`` to eliminate deprecated ``font`` elements. (Thank you, user raiyankamal, for pointing this out.)

--  Slightly more structure is preserved (more empty sublists and strings).

Docx2Python v1 assumed a document was a series of tables and formatted output that way: ``[body[table[table_row[table_cell[paragraph``

Simple docx files *are* structured this way, but there are elements (e.g., ``<w:footnotes>``, ``<w:footnote>``) that act like tables without being exactly tables. Docx2Python v2 treats any element 1-level above a paragraph as a table cell, any element 2-levels above a paragraph as a table row, etc. The upshot of this is that there will be more whitespace in your exports. This whitespace is potentially useful information, but you can easily filter it out if you don't need it.

--  No longer supports Python 3.4, 3.5, or 3.6

Now only supports Python 3.7+

--  XML and other information from an unzipped docx file now available as a DocxReader instance.


docx2python/docx_reader.py
#!/usr/bin/env python3
# _*_ coding: utf-8 _*_
""" Hold and decode docx internal xml files.

:author: Shay Hill
:created: 3/18/2021

See the docx file structure in ``README_DOCX_FILE_STRUCTURE.md``. Each file in that
structure can be stored as a ``File`` instance, though not all will be stored through
the typical docx2python progression.

The ``File`` class is designed to hold and decode xml files with content (text).
Several of the xml files in a docx do not contain content. These contain numbering
formats, font information, rId-lookup tables, and other. ``File`` instances will hold
these as well, though they will not have ``rels`` or ``content`` attributes. Will
return an empty dictionary or empty list if asked.

Some of these non-content files are shared between between . The substance of these
files is accessible through the ``DocxContent`` class. This class holds file
instances and decodes shared non-content in a docx file structure.
"""

from __future__ import annotations

import copy
import os
import pathlib
import zipfile
from contextlib import suppress
from dataclasses import dataclass
from operator import attrgetter
from typing import Dict, List, Optional, Set, Union
from warnings import warn
from pathlib import Path

import xml.etree.ElementTree as etree

from .attribute_register import XML2HTML_FORMATTER
from .docx_context import collect_numFmts, collect_rels
from .docx_text import get_text
from .merge_runs import merge_elems

CONTENT_FILE_TYPES = {"officeDocument", "header", "footer", "footnotes", "endnotes"}


@dataclass
class File:
    """The attribute dict of a file in the docx, plus cached data

    The docx lists internal files in various _rels files. Each will be specified with a
    dict of, e.g.::

        {
            'Id': 'rId8',
            'Type': 'http://schemas.openxmlformats.org/.../relationships/header',
            'Target': 'header1.xml'
        }

    This isn't quite enough to infer the structure of the docx. You'll also need to
    know the directory where each attribute dict was found::

        'dir': 'word/_rels'

    That's the starting point for these instances. Other attributes are inferred or
    created at runtime.
    """

    def __init__(self, context: DocxReader, attribute_dict: Dict[str, str]) -> None:
        """
        Point to container DocxContext instance and store attributes as properties.

        :param context: The DocxContent object holding this instance
        :param attribute_dict: Attributes of this file found in the rels, plus 'dir' as
        described above.
        """
        self.context = context
        self.Id = str(attribute_dict["Id"])
        self.Type = os.path.basename(attribute_dict["Type"])
        self.Target = attribute_dict["Target"]
        self.dir = attribute_dict["dir"]

        # old-style cached_property for Python 3.6 and 3.7 support
        self.__path: Union[None, str] = None
        self.__rels_path: Union[None, str] = None
        self.__rels: Union[None, Dict[str, str]] = None
        self.__root_element: Union[None, etree.Element] = None

    def __repr__(self) -> str:
        """File with self.path"""
        return f"File({self.path})"

    @property
    def path(self) -> str:
        """Infer path/to/xml/file from instance attributes

        :returns: path to xml file

        This will take the information in a file specification (from one of the rels
        files, e.g., {Id:'  ', Type:'  ' Target:'  ', dir:'  '}) and infer a path to
        the specified xml file.

        E.g.,
        from     self.dir = '_rels'       self.Target = 'word/document.xml
                    dirname ''          +       dirname 'word/'
                                        +       filename =   'document.xml'
        return `word/document`

        E.g.,
        from     self.dir = 'word/_rels'       self.Target = 'header1.xml
                    dirname 'word'      +            dirname ''
                                        +       filename =   'header1.xml'
        return `word/header1.xml`
        """
        if self.__path is not None:
            return self.__path

        dirs = [os.path.dirname(x) for x in (self.dir, self.Target)]
        dirname = "/".join([x for x in dirs if x])
        filename = os.path.basename(self.Target)
        self.__path = "/".join([dirname, filename])
        return self.__path

    @property
    def _rels_path(self) -> str:
        """Infer path/to/rels from instance attributes.

        :returns: path to rels (which may not exist)

        Every content file (``document.xml``, ``header1.xml``, ...) will have its own
        ``.rels`` file--if any relationships are defined.

        The path inferred here may not exist.

        E.g.,
        from     self.dir = '_rels'       self.Target = 'word/document.xml
                    dirname ''          +       dirname 'word/'
                                        +       filename =   'document.xml'
        return `word/_rels/document.xml.rels`

        E.g.,
        from     self.dir = 'word/_rels'       self.Target = 'header1.xml
                    dirname 'word'      +            dirname ''
                                        +       filename =   'header1.xml'
        return `word/_rels/header1.xml.rels`
        """
        if self.__rels_path is not None:
            return self.__rels_path
        dirname, filename = os.path.split(self.path)
        self.__rels_path = "/".join([dirname, "_rels", filename + ".rels"])
        return self.__rels_path

    @property
    def rels(self) -> Dict[str, str]:
        """rIds mapped to values

        Each content file.xml will have a file.xml.rels file--if relationships are
        defined. Inside file.xml, values defined in the file.xml.rels file may be
        referenced by their rId numbers.

        :returns: Contents of the file.xml.rels file with reference rId numbers. These
        refer to values defined in the file.xml.rels file:

        E.g.::

        {
            "rId3": "webSettings.xml",
            "rId2": "settings.xml",
            "rId1": "styles.xml",
            "rId6": "theme/theme1.xml",
            "rId5": "fontTable.xml",
            "rId4": "https://www.shayallenhill.com/",
        }

        Not every xml file with have a rels file. Return an empty dictionary if the
        rels file is not found.
        """
        if self.__rels is not None:
            return self.__rels

        try:
            unzipped = self.context.zipf.read(self._rels_path)
            tree = etree.fromstring(unzipped)
            self.__rels = {str(x.attrib["Id"]): str(x.attrib["Target"]) for x in tree}
        except KeyError:
            self.__rels = {}
        return self.__rels

    @property
    def root_element(self) -> etree.Element:
        """Root element of the file.

        Try to merge consecutive, duplicate (except text) elements in content files.
        See documentation for ``merge_elems``. Warn if ``merge_elems`` fails.
        (I don't think it will fail).
        """
        if self.__root_element is not None:
            return self.__root_element

        root = etree.fromstring(self.context.zipf.read(self.path))
        if self.Type in CONTENT_FILE_TYPES:
            root_ = copy.copy(root)
            try:
                merge_elems(self, root)
            except Exception as ex:
                warn(
                    f"Attempt to merge consecutive elements in "
                    f"{self.context.docx_filename} {self.path} resulted in "
                    f"{repr(ex)}. Moving on."
                )
                self.__root_element = root_
        self.__root_element = root
        return self.__root_element

    @property
    def content(self) -> List[List[List[List[str]]]]:
        """Text extracted into a 5-layer-deep nested list of strings."""
        return get_text(self)

    def get_content(
        self, root: Optional[etree.Element] = None
    ) -> List[List[List[List[str]]]]:
        """
        The same content as property 'content' with optional given root.

        :param root: Extract content of file from root down.
            If root is not given, return full content of file.
        """
        return get_text(self, root)


@dataclass
class DocxReader:
    """
    Hold File instances and decode information shared between them (e.g., numFmts)
    """

    def __init__(
        self,
        docx_filename: Union[Path, str],
        html: bool = False,
        paragraph_styles: bool = False,
    ):
        self.docx_filename = docx_filename
        self.do_pStyle = paragraph_styles

        if html:
            self.xml2html_format = XML2HTML_FORMATTER
        else:
            self.xml2html_format = {}

        # old-style cached_propery for Python 3.6 and 3.7 support
        self.__zipf: Union[None, zipfile.ZipFile] = None
        self.__files: Union[None, List[File]] = None
        self.__numId2NumFmts: Union[None, Dict[str, List[str]]] = None

    @property
    def zipf(self) -> zipfile.ZipFile:
        """
        Entire docx unzipped into bytes.
        """
        if self.__zipf is not None:
            return self.__zipf

        self.__zipf = zipfile.ZipFile(self.docx_filename)
        return self.__zipf

    @property
    def files(self) -> List[File]:
        """
        Instantiate a File instance for every content file.
        """
        if self.__files is not None:
            return self.__files

        files = []
        for k, v in collect_rels(self.zipf).items():
            files += [File(self, {**x, "dir": os.path.dirname(k)}) for x in v]
        self.__files = files
        return self.__files

    # noinspection PyPep8Naming
    @property
    def numId2numFmts(self) -> Dict[str, List[str]]:
        """
        numId referenced in xml to list of numFmt per indentation level

        See docstring for collect_numFmts

        Returns an empty dictionary if word/numbering.xml cannot be found.
        Ultimately, this will result in any lists (there should NOT be any lists if
        there is no word/numbering.xml) being "numbered" with "--".
        """
        if self.__numId2NumFmts is not None:
            return self.__numId2NumFmts

        try:
            numFmts_root = etree.fromstring(self.zipf.read("word/numbering.xml"))
            self.__numId2NumFmts = collect_numFmts(numFmts_root)
        except KeyError:
            self.__numId2NumFmts = {}
        return self.__numId2NumFmts

    def file_of_type(self, type_: str) -> File:
        """
        Return file instance attrib Type='http://.../type_'. Warn if more than one.

        :param type_: this package looks for any of
            ("header", "officeDocument", "footer", "footnotes", "endnotes")
            You can try others.
        :return: File instance of the requested type
        """
        files_of_type = self.files_of_type(type_)
        if len(files_of_type) > 1:
            warn("Multiple files of type '{type_}' found. Returning first.")
        try:
            return files_of_type[0]
        except IndexError:
            raise KeyError(
                f"There is no item of type '{type_}' "
                "in the {self.docx_filename} archive"
            )

    def files_of_type(self, type_: Optional[str] = None) -> List[File]:
        """
        File instances with attrib Type='http://.../type_'

        :param type_: this package looks for any of
            ("header", "officeDocument", "footer", "footnotes", "endnotes")
            You can try others. If argument is None (default), returns all content file
            types.
        :return: File instances of the requested type, sorted by path
        """
        if type_ is None:
            types = CONTENT_FILE_TYPES
        else:
            types = {type_}

        return sorted(
            (x for x in self.files if x.Type in types), key=attrgetter("path")
        )

    def content_files(self) -> List[File]:
        """
        Content files (contain displayed text) inside the docx.

        :return: File instances of context files, sorted by path
        """
        return self.files_of_type()

    def save(self, filename: Union[Path, str]) -> None:
        """
        Save the (presumably altered) xml.

        :param filename: path to output file (presumably *.docx)

        xml (root_element) attributes are cached, so these can be altered and saved.
        This allows saving a copy of the input docx after the ``merge_elems`` operation.
        Also allows some light editing like search and replace.
        """
        content_files = [x for x in self.files if x.Type in CONTENT_FILE_TYPES]
        with zipfile.ZipFile(f"{filename}", mode="w") as zout:
            _copy_but(self.zipf, zout, {x.path for x in content_files})
            for file in content_files:
                zout.writestr(file.path, etree.tostring(file.root_element))

    def pull_image_files(
        self, image_directory: Optional[str] = None
    ) -> Dict[str, bytes]:
        """
        Copy images from zip file.

        :param image_directory: optional destination for copied images
        :return: Image names mapped to images in binary format.

            To write these to disc::

                with open(key, 'wb') as file:
                    file.write(value)

        :side effects: Given an optional image_directory, will write the images out
        to file.
        """
        images = {}
        for image in self.files_of_type("image"):
            with suppress(KeyError):
                images[os.path.basename(image.Target)] = self.zipf.read(image.path)
        if image_directory is not None:
            pathlib.Path(image_directory).mkdir(parents=True, exist_ok=True)
            for file, image_bytes in images.items():
                with open(os.path.join(image_directory, file), "wb") as image_copy:
                    image_copy.write(image_bytes)
        return images


def _copy_but(
    in_zip: zipfile.ZipFile, out_zip: zipfile.ZipFile, exclusions: Optional[Set] = None
) -> None:
    """
    Copy every file in a docx except those listed in exclusions.

    :param in_zip: zipfile of origin xml file
    :param out_zip: zipfile of destination xml file
    :param exclusions: filenames you don't want to copy (e.g., 'document.xml')
    """
    exclusions = exclusions or set()
    for item in in_zip.infolist():
        if item.filename not in exclusions:
            buffer = in_zip.read(item.filename)
            out_zip.writestr(item, buffer)
